drain/drain.py

```diff
@@ -0,0 +1,417 @@
+    peer = plugin.rpc.listpeers(peer_id).get('peers')[0]
+    channel = next(c for c in peer['channels'] if 'short_channel_id' in c and c['short_channel_id'] == scid)
+    if check_state:
+        if channel['state'] != "CHANNELD_NORMAL":
+            raise RpcError(payload['command'], payload, {'message': 'Channel %s not in state CHANNELD_NORMAL, but: %s' % (scid, channel['state']) })
+        if not peer['connected']:
+            raise RpcError(payload['command'], payload, {'message': 'Channel %s peer is not connected.' % scid})
+    return channel
+
+
+def spendable_from_scid(plugin, payload, scid):
+    # only fetch funds once to reduce RPC load
+    if not "funds" in payload:
+        payload['funds'] = plugin.rpc.listfunds().get('channels')
+
+    try:
+        channel_funds = next(c for c in payload['funds'] if 'short_channel_id' in c and c['short_channel_id'] == scid)
+    except StopIteration:
+        return Millisatoshi(0), Millisatoshi(0)
+    peer_id = channel_funds['peer_id']
+    funds_our = Millisatoshi(channel_funds['our_amount_msat'])
+    try:
+        channel_peer = get_channel(plugin, payload, peer_id, scid, True)
+    except RpcError:
+        return Millisatoshi(0), Millisatoshi(0)
+
+    # we check amounts via gossip and not wallet funds, as its more accurate
+    our = Millisatoshi(channel_peer['to_us_msat'])
+    total = Millisatoshi(channel_peer['total_msat'])
+    our_reserve = Millisatoshi(channel_peer['our_reserve_msat'])
+    their_reserve = Millisatoshi(channel_peer['their_reserve_msat'])
+    their = total - our
+
+    # reserves maybe not filled up yet
+    if our < our_reserve:
+        our_reserve = our
+    if their < their_reserve:
+        their_reserve = their
+
+    spendable = channel_peer['spendable_msat']
+    receivable = their - their_reserve
+    return spendable, receivable
+
+
+def peer_from_scid(plugin, payload, short_channel_id, my_id):
+    channels = plugin.rpc.listchannels(short_channel_id).get('channels')
+    for ch in channels:
+        if ch['source'] == my_id:
+            return ch['destination']
+    raise RpcError(payload['command'], payload, {'message': 'Cannot find peer for channel: ' + short_channel_id})
+
+
+def find_worst_channel(route):
+    if len(route) < 4:
+        return None
+    start_id = 2
+    worst = route[start_id]['channel']
+    worst_val = route[start_id - 1]['msatoshi'] - route[start_id]['msatoshi']
+    for i in range(start_id + 1, len(route) - 1):
+        val = route[i - 1]['msatoshi'] - route[i]['msatoshi']
+        if val > worst_val:
+            worst = route[i]['channel']
+            worst_val = val
+    return worst
+
+
+def test_or_set_chunks(plugin, payload, my_id):
+    scid = payload['scid']
+    cmd = payload['command']
+    spendable, receivable = spendable_from_scid(plugin, payload, scid)
+    total = spendable + receivable
+    target = total * (0.01 * (100 - payload['percentage']))
+
+    if cmd == "drain":
+        if target >= spendable:
```

Comment:
FWIW after testing if I drain a channel at exactly 50% then these conditions are mutually exclusive ;-)

Ex:
```bash
$ lnreg1 drain 366x1x0 50 50
[
   "987762msat sent over 3 hops to drain 987740msat [1/50]",
   "997392msat sent over 3 hops to drain 997370msat [2/50]",
   "1007783msat sent over 3 hops to drain 1007759msat [3/50]",
   ....

$ lnreg1 drain 366x1x0 1 20
"Error while processing drain: RpcError(\"RPC call failed: ..... error: {'message': 'channel already below target value'}\")"
$ lnreg1 fill 366x1x0 1 20
"Error while processing fill: RpcError(\"RPC call failed: .......  error: {'message': 'channel already above target value'}\")"
```
```

Reply:
What's fun is that rebalanced all the regtest btw ^^

Reply:
Hmm think I'll have to test on the only viable test network : mainnet